massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1074 @@
+        }
+    }
+}
+
+/// A message to update the `MipState`
+#[derive(Clone, Debug)]
+pub struct Advance {
+    /// from VersioningInfo.start
+    pub start_timestamp: MassaTime,
+    /// from VersioningInfo.timeout
+    pub timeout: MassaTime,
+    /// % of past blocks with this version
+    pub threshold: Amount,
+    /// Current time (timestamp)
+    pub now: MassaTime,
+}
+
+// Need Ord / PartialOrd so it is properly sorted in BTreeMap
+
+impl Ord for Advance {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.now).cmp(&other.now)
+    }
+}
+
+impl PartialOrd for Advance {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Advance {
+    fn eq(&self, other: &Self) -> bool {
+        self.start_timestamp == other.start_timestamp
+            && self.timeout == other.timeout
+            && self.threshold == other.threshold
+            && self.now == other.now
+    }
+}
+
+impl Eq for Advance {}
+
+transitions!(MipState,
+    [
+        (Defined, Advance) => [Defined, Started, Failed],
+        (Started, Advance) => [Started, LockedIn, Failed],
+        (LockedIn, Advance) => [LockedIn, Active],
+        (Active, Advance) => Active,
+        (Failed, Advance) => Failed
+    ]
+);
+
+impl Defined {
+    /// Update state from state Defined
+    pub fn on_advance(self, input: Advance) -> MipState {
+        match input.now {
+            n if n > input.timeout => MipState::failed(),
+            n if n > input.start_timestamp => MipState::started(Amount::zero()),
+            _ => MipState::Defined(Defined {}),
+        }
+    }
+}
+
+impl Started {
+    /// Update state from state Started
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            return MipState::failed();
+        }
+
+        if input.threshold >= VERSIONING_THRESHOLD_TRANSITION_ACCEPTED {
+            MipState::locked_in()
+        } else {
+            MipState::started(input.threshold)
+        }
+    }
+}
+
+impl LockedIn {
+    /// Update state from state LockedIn ...
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            MipState::active()
+        } else {
+            MipState::locked_in()
+        }
+    }
+}
+
+impl Active {
+    /// Update state (will always stay in state Active)
+    pub fn on_advance(self, _input: Advance) -> Active {
+        Active {}
+    }
+}
+
+impl Failed {
+    /// Update state (will always stay in state Failed)
+    pub fn on_advance(self, _input: Advance) -> Failed {
+        Failed {}
+    }
+}
+
+/// Wrapper of MipState (in order to keep state history)
+#[derive(Debug, Clone, PartialEq)]
+pub struct MipStateHistory {
+    pub(crate) state: MipState,
+    history: BTreeMap<Advance, MipStateTypeId>,
+}
+
+impl MipStateHistory {
+    /// Create
+    pub fn new(defined: MassaTime) -> Self {
```

Comment:
When creating the history of all the versions, I find it more understandable to have either an empty history or a version state to initialize. I don't fully understand the parameter `defined`